modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php

```diff
@@ -0,0 +1,355 @@
+
+	/**
+	 * On the cart page, add an event listener for removal of product click
+	 */
+	public function remove_from_cart() {
+
+		// We listen at div.woocommerce because the cart 'form' contents get forcibly
+		// updated and subsequent removals from cart would then not have this click
+		// handler attached.
+		$blogid = Jetpack::get_option( 'id' );
+		wc_enqueue_js(
+			"jQuery( 'div.woocommerce' ).on( 'click', 'a.remove', function() {
+				var productID = jQuery( this ).data( 'product_id' );
+				var quantity = jQuery( this ).parent().parent().find( '.qty' ).val()
+				var productDetails = {
+					'id': productID,
+					'quantity': quantity ? quantity : '1',
+				};
+				_wca.push( {
+					'_en': 'woocommerceanalytics_remove_from_cart',
+					'blog_id': '" . esc_js( $blogid ) . "',
+					'pi': productDetails.id,
+					'pq': productDetails.quantity,
+					'ui': '" . esc_js( $this->get_user_id() ) . "',
+				} );
+			} );"
+		);
+	}
+
+	/**
+	 * Adds the product ID to the remove product link (for use by remove_from_cart above) if not present
+	 *
+	 * @param string $url url.
+	 * @param string $key key.
+	 * @return mixed.
+	 */
+	public function remove_from_cart_attributes( $url, $key ) {
+		if ( false !== strpos( $url, 'data-product_id' ) ) {
+			return $url;
+		}
+
+		$item    = WC()->cart->get_cart_item( $key );
+		$product = $item['data'];
+
+		$new_attributes = sprintf(
+			'href="%s" data-product_id="%s" data-product_sku="%s"',
+			esc_attr( $url ),
+			esc_attr( $product->get_id() ),
+			esc_attr( $product->get_sku() )
+		);
+		$url            = str_replace( 'href=', $new_attributes );
+		return $url;
+	}
+
+	/**
+	 * Track a product page view
+	 */
+	public function product_detail() {
+
+		global $product;
+		$blogid = Jetpack::get_option( 'id' );
+
+		$item_details = array(
+			'id'       => $product->get_id(),
+			'name'     => $product->get_title(),
+			'category' => Jetpack_WooCommerce_Analytics_Utils::get_product_categories_concatenated( $product ),
+			'price'    => $product->get_price(),
+		);
+		wc_enqueue_js(
+			"_wca.push( {
+				'_en': 'woocommerceanalytics_product_view',
+				'blog_id': '" . esc_js( $blogid ) . "',
+				'pi': '" . esc_js( $item_details['id'] ) . "',
+				'pn': '" . esc_js( $item_details['name'] ) . "',
+				'pc': '" . esc_js( $item_details['category'] ) . "',
+				'pp': '" . esc_js( $item_details['price'] ) . "',
+				'ui': '" . esc_js( $this->get_user_id() ) . "',
+			} );"
+		);
+	}
+
+	/**
+	 * On the Checkout page, trigger an event for each product in the cart
+	 */
+	public function checkout_process() {
+
+		$universal_commands = array();
+		$cart               = WC()->cart->get_cart();
+		$blogid             = Jetpack::get_option( 'id' );
+
+		foreach ( $cart as $cart_item_key => $cart_item ) {
+			/**
+			* This filter is already documented in woocommerce/templates/cart/cart.php
+			*/
+			$product = apply_filters( 'woocommerce_cart_item_product', $cart_item['data'], $cart_item, $cart_item_key );
```

Comment:
Does this filter need to be used in `product_detail()` above too? If so maybe a helper method `get_item_details()` could be made to dry up this array creation.

Reply:
This filter retrieves the `$product` from the cart, so I don't believe we need it in `product_detail`. However, `get_item_details()` seems like a good idea.